badger/attrstore.go

```diff
@@ -0,0 +1,295 @@
+// Copyright 2017 Pilosa Corp.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package badger
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/cespare/xxhash"
+	"github.com/dgraph-io/badger"
+	"github.com/pilosa/pilosa"
+	"github.com/pkg/errors"
+)
+
+// attrBlockSize is the size of attribute blocks for anti-entropy.
+const attrBlockSize = 100
+
+type attrStore struct {
+	path string
+	db   *badger.DB
+	mu   sync.RWMutex
+}
+
+//NewAttrStore returns a new dgraph-io/badger AttrStore
+func NewAttrStore(path string) pilosa.AttrStore {
+	return newCache(&attrStore{
+		path: path,
+		mu:   sync.RWMutex{},
+	})
+}
+
+// Path returns the path to the store's data file.
+func (s *attrStore) Path() string {
+	return s.path
+}
+
+func (s *attrStore) Open() error {
+	//using default options:
+	opts := badger.DefaultOptions
+	//using a smaller value log file size works
+	//for GOARCH=386
+	opts.ValueLogFileSize = 1 << 29
```

Comment:
There are a couple techniques we could use to provide a different value on 32 bit architectures. It may also be necessary to provide a separate value for running tests (e.g. in CI) which can sometimes be memory constrained. We did something similar for key translation here: https://github.com/pilosa/pilosa/blob/master/translate.go#L105

I'd prefer not to expose another config option (which would be specific to badger) if we can avoid it. It's not exactly clear to me right now what the tradeoffs are for setting this to different values.

Reply:
I haven't finished wrapping my head around this. The default ValueLogFileSize should be small enough for a 32 bit architecture. I also managed to reproduce this locally, this fixed it, but it failed. I am still not certain. However, I wanted to push something up and get feedback.
👍 for not adding additional configuration. I'll add a WIP to the PR.